clusterloader/docs/vision.md

```diff
@@ -0,0 +1,320 @@
+# Cluster loader vision
+
+Author: wojtek-t
+
+Last update time: 9th May 2018
+
+## Background
+
+As of 31/03/2018, all our scalability tests, are regular e2e tests written in Go.
+It makes them really unfriendly for developers not working on scalability who just
+want to load tests Kubernetes features they are working on. Doing so in many
+situation requires understanding how those tests really work, modify their code
+to test the new feature and only then run and debug tests (or create load test for
+your feature on your own, which may be easier, but on the other hand may not
+exercise important metrics that our performance test check). This workflow is far
+from optimal.
+
+That said, long time ago we came up with the idea that users should be able to
+just bring their own objects definitions in JSON format, potentially annotate them
+with some metadata to describe how the load should be generated from these and
+testing infrastructure should do everything else automatically.
+
+In early 2017 the prototype of “Cluster Loader” has been created. It proved that
+configuring tests with json/yaml files is possible. But its functionality is very
+limited and it was very far from enabling migration any of existing scalability
+tests to that framework.
+
+We would like to get back to that and build fully functional Cluster Loader and
+use it as a framework to run all our scalability tests. This doc is describing
+high-level vision of how this will work.
+
+
+## Vision
+
+At the high level, a single test will consist of a number of phases (if we have
+really good usecase we may consider allowing the to create a DAG, but for now
+I don’t have rationale for it). In each phase we will be creating/updating/deleting
+a number of different objects.
+Additionally, we will introduce a set of predefined operations (that user will be
+able to use as phases). They will allow users to monitor/measure performance impact
+of user-defined phases.
+The following subsections will describe this in a bit more details.
+
+
+### Tuning Set
+
+Since we would like to be fully load even very big clusters, we need to be able to
+create a number of “similar” objects. A “Tuning Set” concept will allow us to
+spread those operations across some time.
+We define Tuning Set as following:
+
+```
+struct TuningSet {
+	Name string
+	// Exactly one of the following should be set.
+	QpsLoad        *QpsLoad
+	RandomizedLoad *RandomizedLoad
+	SteppedLoad    *SteppedLoad
+}
+
+// QpsLoad defines a uniform load with a given QPS.
+struct QpsLoad {
+	InitialDelay time.Duration
```

Comment:
I sense that InitialDelay is a characteristic that can be common across the strategies.
Do you think we should move it up into TuningSet?

Reply:
Yes - done.